Three blog posts around TSLint: two on ESLint and one on maintenance (#44)
JoshuaKGoldberg/Goldblog: content/blog/tslint-maintenance/index.md

---
title: "Thoughts as a TSLint Maintainer"
date: "2019-11-18T12:34:56.117Z"
description: "Now that TSLint is being deprecated, let's look at the history of JavaScript and TypeScript linting."
---

> Howdy!
> This blog post is a personal reflection on my experiences working with TSLint.

> > If you're curious about the backing historical context, go back to [TSLint to ESLint Part 1: Historical Drama](../tslint-to-eslint-history).

Author: ```suggestion
> If you're curious about the backing historical context, go back to [TSLint to ESLint Part 1: Historical Context](../tslint-to-eslint-history).
```

_Now_ it's obvious I was wrong.
I was very wrong.
I was incredibly, obviously, plainly wrong.

_Of course_ we would be able to bring the full TypeScript type checker to ESLint.
_Of course_ the TypeScript team would choose to support the open source project already commonplace in the rest of the JavaScript ecosystem.
_Of course_ the relevant maintainer teams would be able to resolve the [assorted](https://github.com/typescript-eslint/typescript-eslint/issues/1132) [structural](https://github.com/typescript-eslint/typescript-eslint/issues/1126) [performance](https://github.com/typescript-eslint/typescript-eslint/issues/1079) **[issues](https://github.com/typescript-eslint/typescript-eslint/issues/1040)**.

> Linking to those issues is not snark or passive-aggressiveness -- it's a celebration of the incredible work done by the TypeScript and typescript-eslint teams in molding together multiple fundamentally different systems.
> We all owe them a _massive_ debt of gratitude for their work to solidify our linting practices.

But, still, in 2018 the commonly accepted way to lint TypeScript code using rules that had access to the type checker was with TSLint.

> Which had those 20 pages of untriaged issues and 70 pull requests left open.

Author: ```suggestion
Which had 20 pages of untriaged issues and 70 pull requests left open.
```

* I got into debates [without explaining my context](https://github.com/palantir/tslint/issues/975#issuecomment-435640297) or [over unnecessary esoteric features](https://github.com/palantir/tslint/issues/1306)...
* I [closed issues prematurely](https://github.com/palantir/tslint/issues/1489#issuecomment-460655323)...
* I [added rules and features without documenting them well](https://github.com/palantir/tslint/issues/4117), causing headaches for beginner and advanced users...
* I [indicated a big refactor would be accepted]() even though we didn't have the maintenance staffing to support it...

Taking on maintenance for a project _(hopefully)_ means you care about where it goes: you want it and its users to be made more productive by the time you spend on it.
I took on a real emotional weight by committing to spending personal time on TSLint.
We can talk all we want about work/life balance and how volunteer work should be a pleasant experience and how open source should be a golden utopia of working towards common goals... but at the end of the day, many projects are free labor provided out of some feeling of commitment and goodwill.

> 

Author: I'd like to add more content to the end of this about the other side. Top of mind is the_"brought to you by: racists"_ event -- how we _could_ have put effort as a community to move TSLint away from Palantir, but flamed around instead.

Author: Ehh, not really relevant to this post - I wasn't a maintainer at the time. What' I'm really getting at is the emotional burden. I'll talk about that.

---
title: "Thoughts as a TSLint Maintainer"
date: "2019-11-18T12:34:56.117Z"
description: "Maintaining TSLint was a wonderful mixed bag of shenanigans. Let's talk about it."
---

> Howdy!
> This blog post is a personal reflection on my experiences working with TSLint.

> > If you're curious about TSLint's history, go back to [TSLint to ESLint Part 1: Historical Context](../tslint-to-eslint-history).

Reviewer: Is it intentionally skips `Migrating from` part form other post titles?

Author: Oh yeah, I'm getting rid of that... nice 👍

No other maintainers were active and the library hadn't been updated in quite a while.
My first goal was to sludge through the hundreds of unaddressed or stale issues and pull requests.

### But Why Bother In The First Place?

Back in 2018 it wasn't clear _(at least to me)_ what was going on with typescript-eslint.
I wasn't sure it would achieve feature parity with TSLint's typed rules in a performant manner to run with ESLint.

_Now_ it's obvious I was wrong.
I was very wrong.
I was incredibly, obviously, plainly wrong.

_Of course_ we would be able to bring the full TypeScript type checker to ESLint.
_Of course_ the TypeScript team would choose to support the open source project already commonplace in the rest of the JavaScript ecosystem.
_Of course_ the relevant maintainer teams would be able to resolve the [assorted](https://github.com/typescript-eslint/typescript-eslint/issues/1132) [structural](https://github.com/typescript-eslint/typescript-eslint/issues/1126) [performance](https://github.com/typescript-eslint/typescript-eslint/issues/1079) **[issues](https://github.com/typescript-eslint/typescript-eslint/issues/1040)**.

> 

Reviewer 2: Tbh, don't quite understand the connotation of the 'of course's in this paragraph. I assume they're sarcastic/ironic since it's non-obvious that this work would've happened successfully, but I'm not sure what that sort of sarcasm adds here.

I'd maybe suggest just making it more straightforward/positive, perhaps:

> But somehow, we were able to bring the full TypeScript type checker...
> And astutely, the TypeScript team would choose...
> And amazingly, the relevant maintainer teams were able to resolve...

Which was a project I'd used to great effect in my personal and work projects -- that I could see had incredible potential for improving developers' TypeScript code.

### What Is and What Should Never Be

My longer-term goal after joining TSLint would have been to move it towards a community management team.
That would have helped buffer TSLint from fluctuating attention on the Palantir side _(though I'm sure the developers there would have continued to play a big part in contributing!)_, and helped bring in the developer power to catch up to ESLint.

We'll never know what that would have been like now.
TypeScript [announced its switch of focus to ESLint](https://github.com/microsoft/TypeScript/issues/29288) a couple of months after I started on TSLint.
As exciting as it was to take on maintaining TSLint, it was _even more exciting_ to see the plans to switch to ESLint.

### Was It Worth It?

_**Yeah!**_
_Of course!_

I wanted to step up and [make the _(linting)_ world a better place](https://www.youtube.com/watch?v=fRUAJVKlUZQ).
I think I objectively did.

> _We_ did -- through plenty of people, including other maintainers and [hundreds of contributors](https://github.com/palantir/tslint/graphs/contributors).

Reviewer 2: Perhaps not relevant for this post, and I've never run the numbers, but I did feel like TSLint got a fairly large amount of new contributors compared to lots of open source projects I've seen. I think, as you mention below, that was due to things like easy to understand code structure (for existing rules), easy to write tests, etc.

Author: Ha indeed. I'll add it to my list of todos.. 😅

Users need to understand the reasons behind maintainer decisions so they can work with the system as they're improving it.

Maintainers ideally set up processes to allow these streams of information to flow freely and share information as necessary.
Users ideally work within those streams to contribute meaningfully.
Ideally, everybody plays nice and understands that both sides want what's best for everyone.

### Being a Maintainer Sucks

Yet, open source software is built over the internet.
Do you know what type of people inhabit the internet?

!["Assholes, assholes everywhere" captioned meme of Buzz and Woody from Toy Story](./assholes-everywhere.png)

<em style="display:block;margin-bottom:2rem;text-align:center;">
This meme usage is inaccurate: it's really that people have the best intentions, but when viewed through the limited lens of the internet, everyone <strong>seems</strong> like an asshole...
</em>

I'll admit I made mistakes and was the "asshole" plenty of times throughout my maintenance tenure:

* I [merged a PR that duplicated another person's work](https://github.com/palantir/tslint/pull/3992#issuecomment-436633059).

> * I got into debates [without explaining my context](https://github.com/palantir/tslint/issues/975#issuecomment-435640297) or [over unnecessary esoteric features](https://github.com/palantir/tslint/issues/1306).

Reviewer 2: I have no regrets for closing that feature request for mixes of tabs and spaces 😆

But don't go overboard -- don't let it be a significant portion of your life unless you're ready to manage significant emotional balancing.

Eventually I'd like to take on more responsibilities in the community again.
But it'll be a while before I'm ready.

> If you haven't yet, read Henry Zhu's [Open Source of Anxiety](https://increment.com/open-source/open-source-of-anxiety) post.
> I only experienced a sliver of burden; maintainers of massive projects such as Babel have _infinitely_ more to deal with.

## Shoutouts

I spent some time looking back through my GitHub activity in the various TSLint-related repositories.
A few names stuck out as consistent supporters of me, the project, and/or the community.

### TSLint

[**Adi Dahiya**](https://github.com/adidahiya) is an obvious MVP here, as the only other active TSLint maintainer in 2019 -- despite switching to studying a completely separate education track!
Adi was involved in the project early on and, through PR reviews, helped me understand how to play in the TSLint arena.
Additional shouting-out belongs here for spot-checking my early work as a maintainer and helping keep me on track while I figured out what I was doing.

> [**Jason Killian**](https://github.com/JKillian) was similarly helpful back in the day, and has since gone on to help run the ever-interesting TypeScriptNYC meetup.

Reviewer 2: might as well link it 😆 

https://www.meetup.com/TypeScriptNYC/

> My [first substantial TSLint PR](https://github.com/palantir/tslint/pulls/1107) back in April of 2016 involved a perfectly informational technical discussion with Jason.

Reviewer 2: link was broken:

```suggestion
My [first substantial TSLint PR](https://github.com/palantir/tslint/pull/1107) back in April of 2016 involved a perfectly informational technical discussion with Jason.
```

[**Klaus Meinhardt**](https://github.com/ajafff) is a tour de force: both by creating the [`tslint-consistent-codestyle`](https://github.com/ajafff/tslint-consistent-codestyle) community ruleset and absurdly useful [`tsutils`](https://github.com/ajafff/tsutils), and the wonderfully crafted and themed [fimbullinter/wotan](https://github.com/fimbullinter/wotan).

> Klaus never failed to give sound technical advice in my (often poorly thought out) TSLint pull requests while still a TSLint maintainer -- and more recently, even in [TypeScript](https://github.com/Microsoft/TypeScript/pulls/29374).

Reviewer 2: ```suggestion
Also was a broken link

Klaus never failed to give sound technical advice in my (often poorly thought out) TSLint pull requests while still a TSLint maintainer -- and more recently, even in [TypeScript](https://github.com/Microsoft/TypeScript/pull/29374).
```

Reviewer: Omg, comment got into suggested change 0_o

I never did end up working with [**Noah Chen**](https://github.com/nchen63) or [**Andy Hanson**](https://github.com/andy-hanson) either, but enjoyed a solid period of helpful PR reviews from them in 2017.
Sometimes I wonder what it would be like to have worked with [**Ashwin R**](https://github.com/ashwinr), who originally created TSLint.

### tslint-microsoft-contrib

[**Hamlet D'Arcy**](https://github.com/HamletDRC) did super work in setting up tslint-microsoft-contrib as the go-to place for non-core TSLint rules, and was kind enough to add me as a contributor earlier in 2018.

> [**Andrii Dieiev**](https://github.com/IllusionMH) came out of nowhere later on and helped support and eventually maintain the library as well, particularly with a bunch of great improvements to its build and release processes.

Reviewer: `and` is twice in the same sentence. I see(?) that they have a bit different relations, but look a bit strange for non-native speaker.

Adi was involved in the project early on and, through PR reviews, helped me understand how to play in the TSLint arena.
Additional shouting-out belongs here for spot-checking my early work as a maintainer and helping keep me on track while I figured out what I was doing.

[**Jason Killian**](https://github.com/JKillian) was similarly helpful back in the day, and has since gone on to help run the ever-interesting [TypeScriptNYC](https://meetup.com/TypeScriptNYC) meetup.
My [first substantial TSLint PR](https://github.com/palantir/tslint/pull/1107) back in April of 2016 involved a perfectly informational technical discussion with Jason.

[**Klaus Meinhardt**](https://github.com/ajafff) is a tour de force: both by creating the [`tslint-consistent-codestyle`](https://github.com/ajafff/tslint-consistent-codestyle) community ruleset and absurdly useful [`tsutils`](https://github.com/ajafff/tsutils), and the wonderfully crafted and themed [fimbullinter/wotan](https://github.com/fimbullinter/wotan).

> Also was a broken link

Reviewer: I think this and next line was unintentionally in suggestion code block

Author: Oh right you mentioned this and I was doing to fix... Then I forgot. 🙃

My [first substantial TSLint PR](https://github.com/palantir/tslint/pull/1107) back in April of 2016 involved a perfectly informational technical discussion with Jason.

[**Klaus Meinhardt**](https://github.com/ajafff) is a tour de force: both by creating the [`tslint-consistent-codestyle`](https://github.com/ajafff/tslint-consistent-codestyle) community ruleset and absurdly useful [`tsutils`](https://github.com/ajafff/tsutils), and the wonderfully crafted and themed [fimbullinter/wotan](https://github.com/fimbullinter/wotan).

> 

Reviewer: nit. This empty line also from suggestion and breaks next sentence into separate paragraph (wasn't initially separated)

Author: _agh_

Author: Thanks again! Fixed.

Reviewer: I know that I'm terrible.
